modules/post/linux/manage/adduser.rb

```diff
@@ -0,0 +1,247 @@
+    groups = datastore['GROUPS'].split
+    groups += ['sudo'] if datastore['SudoMethod'] == 'GROUP'
+    groups = groups.uniq
+    groups_handled = false
+
+    # Check to see that groups exist or fail
+    group_file = read_file('/etc/group').to_s
+    groups_missing = groups.reject { |group| check_group_exists?(group, group_file) }
+
+    unless groups_missing.empty?
+      if datastore['MissingGroups'] == 'ERROR'
+        fail_with(Failure::NotFound, "groups [#{groups_missing.join(' ')}] do not exist on the system")
+      end
+      print_bad("Groups [#{groups_missing.join(' ')}] do not exist on system")
+      if datastore['MissingGroups'] == 'IGNORE'
+        groups -= groups_missing
+        print_good("Removed #{groups_missing.join(' ')} from target groups")
+      end
+    end
+
+    # Check database to see what OS it is. If it meets specific requirements, This can all be done in a single line
+    binary =
+      if datastore['UseraddBinary']
+        datastore['UseraddBinary']
+      elsif command_exists?('useradd')
+        'useradd'
+      elsif command_exists?('adduser')
+        'adduser'
+      else
+        fail_with(Failure::NotFound, 'Cannot find a binary to add a new user cleanly')
+      end
+    os_platform =
+      if session.type == 'meterpreter'
+        sysinfo['OS']
+      elsif active_db? && framework.db.workspace.hosts.where(address: session.session_host) && framework.db.workspace.hosts.where(address: session.session_host).first.os_name
+        host = framework.db.workspace.hosts.where(address: session.session_host).first
+        if host.os_name == 'linux' && host.os_flavor
+          host.os_flavor
+        else
+          host.os_name
+        end
+      else
+        get_sysinfo[:distro]
+      end
+
+    binary_case = if binary =~ /useradd$/
+                    'useradd'
+                  elsif binary =~ /adduser$/
+                    'adduser'
+                  else
+                    binary end
+    case binary_case
+    when 'useradd'
+      print_status("Running on #{os_platform}")
+      print_status('Useradd exists. Using that')
+      case os_platform
+      when /debian|ubuntu|fedora|centos|oracle|redhat|arch|suse|gentoo/i
+        homedirc = datastore['HOME'].empty? ? '--no-create-home' : "--home-dir #{datastore['HOME']}"
+
+        # Since command can add on groups, checking over groups
+        groupadd = ''
+        if datastore['MissingGroups'] == 'CREATE'
+          groupadd = command_exists?('groupadd') ? 'groupadd' : 'addgroup'
+
+          groups_missing.each do |group|
+            d_cmd_exec("#{groupadd} #{group}")
+            print_good("Added #{group} group")
+          end
+        end
+        groupsc = groups.empty? ? '' : "--groups #{groups.join(',')}"
+
+        # Finally run it
+        d_cmd_exec("#{binary} --password #{passwd} #{homedirc} #{groupsc} --shell #{datastore['SHELL']} --no-log-init #{datastore['USERNAME']}".gsub(/ {2,}/, ' '))
```

Comment:
Do we need to single quote these values? i.e. if the user has a `$` in their new password, I believe this will cause issues as it will be evaluated as an expression by `cmd_exec` when it's run on the remote host

```
--password '#{passwd}'
```

And if the user's provided details have quotes etc that will be problematic

Reply:
I thought that it was working when I was testing, but I just tested again with some containers, and youre absolutely right. Fixed this issue in [761f91f](https://github.com/rapid7/metasploit-framework/pull/18194/commits/761f91faf01878e5739f5f50937167aaf6903f9d).